File: PointFreeLeetCode/PointFree.LeetCode.Tests_XUnit/Problem_1624.fs
module PointFree.LeetCode.Tests_XUnit.Problem_1624

open System.Text.RegularExpressions
open FsCheck.FSharp
open FsCheck.Xunit

open PointFree.LeetCode
open PointFree.LeetCode_CSharp

type NotEmptyLowerCaseString =
    
    static member String() =
        ArbMap.defaults
        |> ArbMap.arbitrary<string>
        |> Arb.mapFilter (_.ToLower()) (fun s -> Regex.IsMatch(s, "^[a-z]+$") && s.Length >= 1 && s.Length <= 300)
    


[<Property(Arbitrary = [| typeof<NotEmptyLowerCaseString> |])>]
let ``Implementation is correct`` (string) =
    let s = Problem_1624()
    
    let r1 = s.MaxLengthBetweenEqualCharacters(string)
    let r2 = Problem_1624.maxLengthBetweenEqualCharacters string
    
    r1 = r2 
    

File: PointFreeLeetCode/PointFree.LeetCode_CSharp/Problem_1624.cs
namespace PointFree.LeetCode_CSharp;

public class Problem_1624
{
    public int MaxLengthBetweenEqualCharacters(string s)
    {
        var cache = new Dictionary<char, int>();
        var res = -1;

        for (var i = 0; i < s.Length; i++)
        {
            if (cache.TryGetValue(s[i], out var val))
            {
                res = Math.Max(res, i - val - 1);
            }
            else
            {
                cache.Add(s[i], i);
            }
        }

        return res;
    }
}